import sys
import os
import base64
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import grpc
import medical_ai_pb2 as pb2
import medical_ai_pb2_grpc as pb2_grpc

def image_to_base64(image_path: str) -> str:
    """å°†å›¾ç‰‡è½¬æ¢ä¸º Base64 ç¼–ç å­—ç¬¦ä¸²"""
    with open(image_path, "rb") as image_file:
        encoded_string = base64.b64encode(image_file.read()).decode('utf-8')
    return f"data:image/png;base64,{encoded_string}"

def get_real_request_data(stream_mode: bool, patient_dept: str):
    """ä½¿ç”¨çœŸå®çš„æ‚£è€…æ•°æ®å’Œå›¾ç‰‡"""
    # çœŸå®æ‚£è€…æ–‡æœ¬æ•°æ®
    patient_text_data = """
    **æ‚£è€…åŸºæœ¬ä¿¡æ¯**
    å§“åï¼šå¼ åŒå­¦
    æ€§åˆ«ï¼šç”·
    å¹´é¾„ï¼š20 å²
    èŒä¸šï¼šåœ¨æ ¡å¤§å­¦ç”Ÿ
    èº«é«˜ï¼š176 cm
    ä½“é‡ï¼š85 kg
    BMIæŒ‡æ•°ï¼šçº¦ 27.4
    è¿‡æ•å²ï¼šæ— 
    æ—¢å¾€ç—…å²ï¼šæ— ç‰¹æ®Šè¯´æ˜

    **ä¸»è¯‰ä¸ç°ç—…å²**
    ä¸»è¯‰ï¼šå’½å–‰å‰§çƒˆç–¼ç—›ã€‚
    ç—‡çŠ¶æè¿°ï¼šæ‚£è€…è‡ªè¿°å–‰å’™éš¾å—ï¼Œä¼´æœ‰æ˜æ˜¾çš„ç—›æ„Ÿã€‚
    """
    
    # è¯»å–çœŸå®å›¾ç‰‡
    image_path = os.path.join(os.path.dirname(__file__), "pic", "tongue_sample.png")
    if not os.path.exists(image_path):
        print(f"âŒ é”™è¯¯ï¼šå›¾ç‰‡æ–‡ä»¶ä¸å­˜åœ¨ {image_path}")
        raise FileNotFoundError(f"å›¾ç‰‡æ–‡ä»¶ä¸å­˜åœ¨: {image_path}")
    
    image_base64 = image_to_base64(image_path)
    print(f"âœ… å·²åŠ è½½å›¾ç‰‡ï¼š{image_path} (Base64é•¿åº¦: {len(image_base64)} å­—ç¬¦)")
    
    return pb2.AnalysisRequest(
        patient_text_data=patient_text_data,
        image_base64=image_base64,
        stream=stream_mode,
        patient_department=patient_dept
    )

def sync_call(patient_dept: str):
    print("\n" + "="*50)
    print(f"ã€åŒæ­¥è°ƒç”¨æµ‹è¯•ã€‘ç—…äººé€‰æ‹©ç§‘å®¤ï¼š{patient_dept}")
    print("="*50)
    
    with grpc.insecure_channel('127.0.0.1:50051') as channel:
        stub = pb2_grpc.MedicalAIServiceStub(channel)
        request = get_real_request_data(stream_mode=False, patient_dept=patient_dept)
        response_iterator = stub.ProcessMedicalAnalysis(request)
        
        for response_chunk in response_iterator:
            if response_chunk.is_end:
                # åŒæ­¥ï¼šbytesâ†’Protobufå¯¹è±¡ï¼ˆååºåˆ—åŒ–ï¼‰
                sync_report = pb2.AnalysisReport.FromString(response_chunk.chunk_data)
                if sync_report.status == "SUCCESS":
                    # æŠ¥å‘Šæ˜¯UTF-8å­—ç¬¦ä¸²ï¼Œç›´æ¥æ‰“å°
                    print("âœ… åŒæ­¥è°ƒç”¨æˆåŠŸï¼Œå®Œæ•´æŠ¥å‘Šï¼š")
                    print(sync_report.structured_report)
                else:
                    print(f"âŒ åŒæ­¥è°ƒç”¨å¤±è´¥")
                    print(f"   çŠ¶æ€: {sync_report.status}")
                    print(f"   æ¶ˆæ¯: {sync_report.message}")
                    if sync_report.structured_report:
                        print(f"   è¯¦æƒ…: {sync_report.structured_report}")
                break

def stream_call(patient_dept: str):
    print("\n" + "="*50)
    print(f"ã€æµå¼è°ƒç”¨æµ‹è¯•ã€‘ç—…äººé€‰æ‹©ç§‘å®¤ï¼š{patient_dept}")
    print("="*50)
    
    with grpc.insecure_channel('127.0.0.1:50051') as channel:
        stub = pb2_grpc.MedicalAIServiceStub(channel)
        request = get_real_request_data(stream_mode=True, patient_dept=patient_dept)
        response_iterator = stub.ProcessMedicalAnalysis(request)
        
        print("ğŸ“¥ å¼€å§‹æ¥æ”¶æµå¼æ•°æ®ï¼š")
        for response_chunk in response_iterator:
            # æµå¼ï¼šbytesâ†’UTF-8å­—ç¬¦ä¸²
            chunk_str = response_chunk.chunk_data.decode('utf-8')
            if chunk_str == "[STREAM_END]":
                print("\nğŸ“¥ æµå¼æ¥æ”¶å®Œæ¯•ï¼ˆæ”¶åˆ°ç»“æŸæ ‡è®°ï¼‰")
                break
            print(f"[æµå¼chunk] {chunk_str}", end="")
            if not response_chunk.is_end:
                print("\n" + "-"*30)

if __name__ == "__main__":
    print("\n" + "="*80)
    print("ğŸŒŸ åŒ»ç–—AIæœåŠ¡ gRPC å®¢æˆ·ç«¯æµ‹è¯•ï¼ˆä½¿ç”¨çœŸå®æ•°æ®ï¼‰")
    print("="*80)
    
    # æµ‹è¯• 1: åŒæ­¥è°ƒç”¨ + æ­£ç¡®ç§‘å®¤
    sync_call(patient_dept="è€³é¼»å–‰ç§‘")
    
    # æµ‹è¯• 2: æµå¼è°ƒç”¨ + æ­£ç¡®ç§‘å®¤
    # stream_call(patient_dept="è€³é¼»å–‰ç§‘")
    
    # æµ‹è¯• 3: åŒæ­¥è°ƒç”¨ + é”™è¯¯ç§‘å®¤
    # sync_call(patient_dept="å†…ç§‘")
    
    # æµ‹è¯• 4: æµå¼è°ƒç”¨ + é”™è¯¯ç§‘å®¤
    # stream_call(patient_dept="å†…ç§‘")
    
    print("\n" + "="*80)
    print("âœ… æµ‹è¯•å®Œæˆ")
    print("="*80)